File: crates/sc-watch/src/hotset.rs
//! Hot-set bookkeeping: which `(share, dir)` keys currently have a live OS
//! watch, split into "sticky" (WebSocket-subscribed dirs + their ancestor
//! chain, plus share roots — never auto-evicted) and "recent" (an LRU of
//! everything else that's been touched, capped at `hot_set_max` and evicted
//! oldest-first —).

use std::collections::HashMap;
use std::num::NonZeroUsize;

use sc_vfs::ShareId;
use lru::LruCache;

pub(crate) type Key = (ShareId, String);

pub(crate) struct HotSet {
    /// Key -> subscriber refcount. Present here ⇒ never evicted by `touch`.
    sticky: HashMap<Key, u32>,
    /// Recently-touched, evictable keys, oldest at the back.
    recent: LruCache<Key, ()>,
    /// Every key that currently has a live backend registration, sticky or
    /// recent — this is the thing capped at `hot_set_max`.
    registered: HashMap<Key, ()>,
    cap: usize,
}

impl HotSet {
    pub(crate) fn new(cap: usize) -> Self {
        Self {
            sticky: HashMap::new(),
            // `recent`'s own capacity is deliberately *not* `cap`: eviction
            // is enforced exclusively through `keys_to_evict_for_one_more`,
            // which also unregisters the backend watch and updates
            // `registered`. If `LruCache` itself were capped at `cap`, its
            // internal `put` would silently self-evict entries the moment
            // `recent.len() == cap` — even while `sticky` entries mean the
            // *real* registered count is still under `cap` — leaving a
            // "ghost" backend watch nothing ever unregisters. Giving it
            // a generous, effectively-unbounded capacity here makes it a
            // pure ordering structure; `keys_to_evict_for_one_more` is the
            // only thing that actually removes entries from it.
            recent: LruCache::new(NonZeroUsize::new(cap.max(1).saturating_add(1_000_000)).unwrap()),
            registered: HashMap::new(),
            cap,
        }
    }

    pub(crate) fn registered_count(&self) -> usize {
        self.registered.len()
    }

    /// Snapshot of every key with a live backend registration. Used by the
    /// periodic NFS/FUSE rescan, which only ever
    /// revisits directories already being tracked — never a fresh
    /// recursive walk.
    pub(crate) fn registered_keys(&self) -> Vec<Key> {
        self.registered.keys().cloned().collect()
    }


    pub(crate) fn mark_registered(&mut self, key: Key) {
        self.registered.insert(key, ());
    }

    pub(crate) fn mark_unregistered(&mut self, key: &Key) {
        self.registered.remove(key);
    }

    /// Add a subscriber to `key`, promoting it out of the recent LRU into
    /// the sticky set. Returns `true` if this is the key's first
    /// registration (caller must register it with the backend).
    pub(crate) fn add_sticky(&mut self, key: Key) -> bool {
        self.recent.pop(&key);
        *self.sticky.entry(key.clone()).or_insert(0) += 1;
        !self.registered.contains_key(&key)
    }

    /// Remove one subscriber from `key`. Once the refcount hits zero it
    /// drops back into the recent LRU (still registered, now evictable)
    /// rather than being torn down immediately.
    pub(crate) fn remove_sticky(&mut self, key: &Key) {
        if let Some(count) = self.sticky.get_mut(key) {
            *count -= 1;
            if *count == 0 {
                self.sticky.remove(key);
                self.recent.put(key.clone(), ());
            }
        }
    }

    /// LRU-bump `key`. Returns `true` if this is a brand new registration
    /// the caller must register with the backend.
    pub(crate) fn touch(&mut self, key: Key) -> bool {
        if self.sticky.contains_key(&key) {
            return false; // already registered and pinned
        }
        let is_new = !self.registered.contains_key(&key);
        self.recent.put(key, ());
        is_new
    }

    /// Keys to evict (oldest-first, non-sticky) so `registered_count() + 1
    /// <= cap` before adding one more registration. Does *not* mutate
    /// `registered`/`recent` itself — the caller evicts from the backend
    /// first (best-effort), then calls `mark_unregistered` for each.
    pub(crate) fn keys_to_evict_for_one_more(&mut self) -> Vec<Key> {
        let mut evicted = Vec::new();
        while self.registered.len() + 1 > self.cap {
            match self.recent.pop_lru() {
                Some((k, _)) => evicted.push(k),
                None => break, // nothing left to evict; let it exceed cap slightly
            }
        }
        evicted
    }
}

File: go/engine/service/watch/hotset.go
package watch

import (
	"container/list"

	"github.com/heavycaffeiner/stowcloud/go/engine/infra/vfs"
)

// key identifies a watched directory relative to its share. Host paths never
// appear here; only the transport requires one.
type key struct {
	share vfs.ShareID
	dir   string
}

// hotSet tracks which directories presently hold a kernel watch. It has two
// halves, and only one behaves as an LRU.
//
// The recent half is bounded and evicts, which is the portion that resembles
// the whole feature. The pinned half is reference counted, never evicted
// automatically, and driven externally: subscribing pins the directory being
// watched and unsubscribing releases it.
//
// Implementing only the LRU half fails silently. Under load an LRU evicts
// exactly the directory a user is currently viewing, so invalidations stop
// reaching the folder in front of them while everything else continues
// working.
type hotSet struct {
	cap int

	// sticky associates each pinned key with how many subscribers hold it.
	sticky map[key]uint32

	// order lists keys by recency with the newest first, and recent indexes into
	// it. Its length does not enforce the cap: eviction runs solely through
	// evictFor, which also unregisters the kernel watch. A self-evicting
	// structure would strand watches that nothing ever removes.
	order  *list.List
	recent map[key]*list.Element

	// registered holds every key with a live watch, pinned or recent. The cap
	// applies to this set.
	registered map[key]struct{}
}

func newHotSet(capacity int) *hotSet {
	capacity = max(capacity, 1)
	return &hotSet{
		cap:        capacity,
		sticky:     make(map[key]uint32),
		order:      list.New(),
		recent:     make(map[key]*list.Element),
		registered: make(map[key]struct{}),
	}
}

func (h *hotSet) registeredCount() int { return len(h.registered) }

// stickyCount is how many keys a subscriber holds pinned.
func (h *hotSet) stickyCount() int { return len(h.sticky) }

func (h *hotSet) isRegistered(k key) bool {
	_, ok := h.registered[k]
	return ok
}

func (h *hotSet) markRegistered(k key)   { h.registered[k] = struct{}{} }
func (h *hotSet) markUnregistered(k key) { delete(h.registered, k) }

// registeredKeys returns a snapshot for the periodic rescan, which revisits
// only already-tracked directories and never walks a tree.
func (h *hotSet) registeredKeys() []key {
	out := make([]key, 0, len(h.registered))
	for k := range h.registered {
		out = append(out, k)
	}
	return out
}

// addSticky pins k, lifting it out of the recent half. The return value says
// whether the caller must now register a kernel watch.
func (h *hotSet) addSticky(k key) bool {
	h.dropRecent(k)
	h.sticky[k]++
	return !h.isRegistered(k)
}

// removeSticky drops one subscriber. Reaching zero returns the key to the
// recent half, still registered but now evictable, rather than tearing it down
// while another reader might be a moment behind.
func (h *hotSet) removeSticky(k key) {
	n, ok := h.sticky[k]
	if !ok {
		return
	}
	n--
	if n > 0 {
		h.sticky[k] = n
		return
	}
	delete(h.sticky, k)
	if h.isRegistered(k) {
		h.pushRecent(k)
	}
}

// touch refreshes k's recency and reports whether the caller must register a
// kernel watch. Pinned keys are untouched: they already sit in the half that
// never evicts, and relocating one would return it to the evictable set.
func (h *hotSet) touch(k key) bool {
	if _, pinned := h.sticky[k]; pinned {
		return false
	}
	isNew := !h.isRegistered(k)
	h.pushRecent(k)
	return isNew
}

// evictFor selects keys to unregister, oldest first, making room for one more
// registration under the cap. It does not remove the kernel watch itself: the
// caller does that and then calls markUnregistered, so a failed removal cannot
// leave the two views inconsistent.
func (h *hotSet) evictFor(k key) []key {
	if h.isRegistered(k) {
		return nil
	}
	var out []key
	for len(h.registered)+1 > h.cap {
		back := h.order.Back()
		if back == nil {
			// Only pinned keys remain. Slightly overshooting the cap is
			// preferable to discarding a watch a user is actively reading.
			break
		}
		evicted, ok := back.Value.(key)
		if !ok {
			h.order.Remove(back)
			continue
		}
		h.dropRecent(evicted)
		out = append(out, evicted)
		// Unregistering is the caller's job; counting it here terminates the
		// loop.
		delete(h.registered, evicted)
	}
	return out
}

func (h *hotSet) pushRecent(k key) {
	if e, ok := h.recent[k]; ok {
		h.order.MoveToFront(e)
		return
	}
	h.recent[k] = h.order.PushFront(k)
}

func (h *hotSet) dropRecent(k key) {
	if e, ok := h.recent[k]; ok {
		h.order.Remove(e)
		delete(h.recent, k)
	}
}
